exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/AgeFunction.java

```diff
@@ -0,0 +1,106 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+<@pp.dropOutputFile />
+<#assign className="GAge"/>
+
+<@pp.changeOutputFile name="/org/apache/drill/exec/expr/fn/impl/${className}.java"/>
+
+<#include "/@includes/license.ftl"/>
+
+package org.apache.drill.exec.expr.fn.impl;
+
+import org.apache.drill.exec.expr.DrillSimpleFunc;
+import org.apache.drill.exec.expr.annotations.FunctionTemplate;
+import org.apache.drill.exec.expr.annotations.FunctionTemplate.NullHandling;
+import javax.inject.Inject;
+import org.apache.drill.exec.expr.annotations.Output;
+import org.apache.drill.exec.expr.annotations.Workspace;
+import org.apache.drill.exec.expr.annotations.Param;
+import org.apache.drill.exec.expr.holders.*;
+import org.apache.drill.exec.record.RecordBatch;
+import org.apache.drill.exec.ops.ContextInformation;
+
+/*
+ * This class is generated using freemarker and the ${.template_name} template.
+ */
+
+public class ${className} {
+
+<#list dateIntervalFunc.dates as fromUnit>
+<#list dateIntervalFunc.dates as toUnit>
+
+  @FunctionTemplate(name = "age",
+                    scope = FunctionTemplate.FunctionScope.SIMPLE,
+                    nulls = FunctionTemplate.NullHandling.NULL_IF_NULL)
+  public static class Age${fromUnit}To${toUnit} implements DrillSimpleFunc {
+
+    @Param ${toUnit}Holder left;
+    @Param ${fromUnit}Holder right;
+    @Output IntervalHolder out;
+
+    public void setup() {
+    }
+
+    public void eval() {
+      java.time.LocalDateTime from = java.time.Instant.ofEpochMilli(right.value).atZone(java.time.ZoneOffset.UTC).toLocalDateTime();
+      java.time.LocalDateTime to = java.time.Instant.ofEpochMilli(left.value).atZone(java.time.ZoneOffset.UTC).toLocalDateTime();
+
+      long months = from.until(to, java.time.temporal.ChronoUnit.MONTHS);
+      from = from.plusMonths(months);
+      long days = from.until(to, java.time.temporal.ChronoUnit.DAYS);
+      from = from.plusDays(days);
+      long millis = from.until(to, java.time.temporal.ChronoUnit.MILLIS);
+
+      out.months = (int) months;
+      out.days = (int) days;
+      out.milliseconds = (int) millis;
+    }
+  }
+</#list>
+  @FunctionTemplate(name = "age",
+                    scope = FunctionTemplate.FunctionScope.SIMPLE,
+                    nulls = FunctionTemplate.NullHandling.NULL_IF_NULL)
+  public static class Age${fromUnit}ToMidnight implements DrillSimpleFunc {
+
+    @Param ${fromUnit}Holder right;
+    @Workspace java.time.LocalDateTime to;
+    @Output IntervalHolder out;
+    @Inject ContextInformation contextInfo;
+
+    public void setup() {
+      java.time.ZoneId zoneId = contextInfo.getRootFragmentTimeZone();
+      java.time.ZonedDateTime zdtStart = contextInfo.getQueryStartInstant().atZone(zoneId);
+      to = zdtStart.truncatedTo(java.time.temporal.ChronoUnit.DAYS).toLocalDateTime();
```

Comment:
Interestingly, here we're using an actual UTC-based Unix timestamp, so the conversion used earlier is the correct one. But, here, are we doing the right thing? We want a local date/time with the end date, as converted from a UTC timestamp. Instead, it seems we are converting from local time to UTC.

Are we trying to do the age calcs in UTC? We should do them in local time so that a baby born at 11:00PM local time yesterday (the data value) will have an age of 2 hours for a query run at 1AM today.

(This is where that definition comment would be handy: the common convention is that the baby is one day old at midnight, though in our terms, at midnight the baby is one hour old. Thus, in our convention, the person won't turn 21, say, until 11PM on their 21th birthday, though a bartender will happily serve them any time on that day.)

A comment to explain this will help future developers.

Reply:
Yes here is indeed a real Unix timestamp.  Actually, other than that, UTC does _not_ feature anywhere here.  Instead we convert this query start time to the the local time zone (`getRootFragmentTimeZone`) get to a LocalDateTime.  The age calcs all happen with LocalDateTimes, not UTC or any time zone, as they do in e.g. `TIMESTAMPDIFF` (including before I refactored it).  If we want to start using zoned date time objects for these calcs I think we should open a discussion about consistently applying that across all functions (so at least also `TIMESTAMPDIFF`) in a separate PR.

At 1am local time, a baby born yesterday at 11pm is reported as being one hour old by the unary form of `AGE` (substract from midnight) and two hours old by the binary invocation `select AGE(current_timestamp, <baby's birth time>`.

I've added comments explaining that AGE does not count calendar boundary crossings, like midnight, as the barman does.